### PR TITLE
Mappings refactoring

### PR DESCRIPTION
**Pull request by vlukes**
This PR removes `refmaps.c`, `mappings.pyx,` and `mappings.pxd`. `CMapping` from `mappings.pyx` is replaced by class `common/mappings.DMappings` and  by function  `fen/mappings.eval_mapping_data_in_qp()`. The interface to term C functions is realized by `DMappings.cmap` which is C data container defined in `cmapping.pyx`/`cmapping.pxd` and `refmaps.h`.

### CODE REVIEW

**Comment by vlukes** [comment]
I will add some docstrings.

File: sfepy/discrete/fem/mappings.py [hunk 1]
```diff
@@ -71,84 +193,85 @@ def get_physical_qps(self, qp_coors):
...
 
+        Parameters
+        ----------
+        qp_coors: numpy.ndarray
+            The coordinates of the integration points.
+        weights:
+            The integration weights.
+        poly_space: PolySpace instance
+            The PolySpace instance.
+        ori: ?
```
Comment (vlukes) [inline sfepy/discrete/fem/mappings.py:210]: @rc, please, could you explain `ori` and `transform`?

File: sfepy/discrete/common/mappings.py [hunk 2]
```diff
@@ -4,6 +4,41 @@
...
+    Data for C functions translated to FMFields and embedded in CMapping.
+    """
+    def __init__(self, bf, det, volume, bfg, normal, dim):
+        self.bf = bf
+        self.det = det
+        self.volume = volume
+        self.bfg = bfg
+        self.normal = normal
+
+        self.cmap = CMapping(bf, det, volume, bfg, normal, dim)
```
Comment (rc) [inline sfepy/discrete/common/mappings.py:22]: So PyCMapping instance is called geo, and we have geo.cmap.geo. Food for thought... maybe something clearer might be devised. The lowest level cmap.geo is dictated by the Cython/C terms code. What about the upper geo? :)

File: sfepy/discrete/fem/mappings.py [hunk 3]
```diff
@@ -5,11 +5,122 @@
 
 from sfepy import Config
 from sfepy.base.base import get_default, output
-from sfepy.base.mem_usage import raise_if_too_large
```
Comment (rc) [inline sfepy/discrete/fem/mappings.py:8]: I would like to preserve this. It allows an early failure when memory will not be sufficient for term evaluations -> infinite swapping.

File: sfepy/discrete/fem/mappings.py [hunk 4]
```diff
@@ -5,11 +5,122 @@
...
+    conn: numpy.ndarray
+        The element connectivity.
+    dim: int
+        The space dimension.
+    n_ep: int
+        The number of element points.
+    bf_g: numpy.ndarray
+        The derivatives of the domain base functions with respect to the
+        reference coordinates.
+    weight: numpy.ndarray
```
Comment (rc) [inline sfepy/discrete/fem/mappings.py:31]: FWIW: While both coor and coors are used, this is weights in Python code.

File: sfepy/discrete/fem/mappings.py [hunk 5]
```diff
@@ -5,11 +5,122 @@
...
+        The weights of the integration points.
```
Comment (rc) [inline sfepy/discrete/fem/mappings.py:32]: I would use quadrature instead of integration.

File: sfepy/discrete/fem/mappings.py [hunk 6]
```diff
@@ -71,84 +192,85 @@ def get_physical_qps(self, qp_coors):
...
+              - the boundary connectivity
+              - the derivatives of the domain boundary base functions with
+                respect to the reference coordinates
+
         Returns
         -------
-        cmap : CMapping instance
-            The volume mapping.
+        pycmap: PyCMapping instance
+            The domain mapping fields.
```
Comment (rc) [inline sfepy/discrete/fem/mappings.py:226]: 'The domain mapping data.' to avoid mistaking it with fields?

File: sfepy/terms/terms_biot.py [hunk 7]
```diff
@@ -150,7 +150,8 @@ def function(out, val_qp, mat, vg, fmode):
             status = 0
 
         else:
-            status = terms.de_cauchy_stress(out, val_qp, mat, vg, fmode)
+
+            status = terms.de_cauchy_stress(out, val_qp, mat, vg.cmap, fmode)
```
Comment (rc) [inline sfepy/terms/terms_biot.py:154]: Cannot Term.get_mapping() return directly vg.cmap as before this PR?

**Comment by rc** [comment]
Thanks, well done! The tests pass but I have a suspect - hanging nodes:

```
python3 sfepy/examples/diffusion/laplace_refine_interactive.py output
...
sfepy: nls: iter: 0, residual: 7.827758e+05 (rel: 1.000000e+00)
sfepy:   residual:    0.38 [s]
sfepy:     matrix:    0.00 [s]
sfepy:      solve:    0.00 [s]
sfepy: warning: linear system solution precision is lower
sfepy: then the value set in solver options! (err = 3.807985e-10 < 1.000000e-10)
sfepy: nls: iter: 1, residual: 4.133822e-10 (rel: 5.280978e-16)
sfepy: solved in 1 steps in 0.39 seconds
```
while the original code leads to:
```
sfepy: nls: iter: 0, residual: 4.785582e+00 (rel: 1.000000e+00)
sfepy:   residual:    0.01 [s]
sfepy:     matrix:    0.00 [s]
sfepy:      solve:    0.00 [s]
sfepy: nls: iter: 1, residual: 2.571117e-15 (rel: 5.372632e-16)
sfepy: solved in 1 steps in 0.01 seconds
```
Notice the residual times and the initial residual magnitudes.

**Comment by rc** [comment]
The initial residual calculation is much slower now (see e.g. the navier-stokes example) - can this be improved?

File: sfepy/discrete/fem/mappings.py (hunk 3 above)
Comment (vlukes) [reply sfepy/discrete/fem/mappings.py:8]: OK, I thought it was only related to allocation i C. And now nothing is allocated in C.

File: sfepy/discrete/fem/mappings.py (hunk 6 above)
Comment (vlukes) [reply sfepy/discrete/fem/mappings.py:226]: Yes, `mapping data` it is better.

File: sfepy/terms/terms_biot.py (hunk 7 above)
Comment (vlukes) [reply sfepy/terms/terms_biot.py:154]: The Python structure is argument of the Python terms, the C structure enters the C terms.  And, in general, I can not determine which function will be called at the time `Term.get_mapping()` is called.

Comment (rc) [reply sfepy/terms/terms_biot.py:154]: OK, thanks.

**Comment by vlukes** [comment]
> Notice the residual times and the initial residual magnitudes.

Fixed!

**Comment by vlukes** [comment]
> The initial residual calculation is much slower now (see e.g. the navier-stokes example) - can this be improved?

Yes, this can be improved using `invs_fast`.

**Comment by rc** [comment]
Thanks! Even with the `*_fast()` functions, it is about 0.2 vs. 0.03 with `sfepy-run sfepy/examples/navier_stokes/navier_stokes.py`. I wonder why the old code was so fast. :)

**Comment by vlukes** [comment]
> Thanks, well done! The tests pass but I have a suspect - hanging nodes:
> 
> ```
> python3 sfepy/examples/diffusion/laplace_refine_interactive.py output
> ...
> sfepy: nls: iter: 0, residual: 7.827758e+05 (rel: 1.000000e+00)
> sfepy:   residual:    0.38 [s]
> sfepy:     matrix:    0.00 [s]
> sfepy:      solve:    0.00 [s]
> sfepy: warning: linear system solution precision is lower
> sfepy: then the value set in solver options! (err = 3.807985e-10 < 1.000000e-10)
> sfepy: nls: iter: 1, residual: 4.133822e-10 (rel: 5.280978e-16)
> sfepy: solved in 1 steps in 0.39 seconds
> ```
> 
> while the original code leads to:
> 
> ```
> sfepy: nls: iter: 0, residual: 4.785582e+00 (rel: 1.000000e+00)
> sfepy:   residual:    0.01 [s]
> sfepy:     matrix:    0.00 [s]
> sfepy:      solve:    0.00 [s]
> sfepy: nls: iter: 1, residual: 2.571117e-15 (rel: 5.372632e-16)
> sfepy: solved in 1 steps in 0.01 seconds
> ```
> 
> Notice the residual times and the initial residual magnitudes.

This is fixed by [fix 2x2 inversion](https://github.com/sfepy/sfepy/pull/937/commits/755c10998f2200f508c3f5e96fbea1ffbf709ce5).

**Comment by rc** [comment]
What about replacing the Python `dets_fast()` by a Cython version, that could be as fast as the original `geme_det3x3()` C function? Maybe another PR?

**Comment by vlukes** [comment]
> What about replacing the Python `dets_fast()` by a Cython version, that could be as fast as the original `geme_det3x3()` C function? Maybe another PR? Actually are we sure that is the sole culprit of the slow-down? I will make some measurements when time permits.

According to my measurement, the Python `dets_fast()` and  `invs_fast()` are fast enough. There is a higher overhead in the Python implementation of `eval_mapping_data_in_qp()` compared to the original C code.

**Comment by rc** [comment]
OK, so it;s the einsum calls and/or coors expansion to connectivity. Let's keep it as it is then, we can try optimizing later, if need arises.

**Comment by rc** [comment]
FWIW and for the record: The `python3 sfepy/examples/diffusion/laplace_refine_interactive.py output` now works, but I have noticed earlier (*) and this confirms the observation that einsum-based evaluations have different round-off error behavior than our original C functions:
- new (einsum):
```
sfepy: nls: iter: 0, residual: 4.785582e+00 (rel: 1.000000e+00)
sfepy:   residual:    0.01 [s]
sfepy:     matrix:    0.00 [s]
sfepy:      solve:    0.00 [s]
sfepy: nls: iter: 1, residual: 2.746520e-15 (rel: 5.739156e-16)
```
- original (C):
```
sfepy: nls: iter: 0, residual: 4.785582e+00 (rel: 1.000000e+00)
sfepy:   residual:    0.01 [s]
sfepy:     matrix:    0.00 [s]
sfepy:      solve:    0.00 [s]
sfepy: nls: iter: 1, residual: 2.571117e-15 (rel: 5.372632e-16)
```
Let us keep that in mind.

(*) E.g. the symmetric `de_*` based matrices are slightly less machine-precision-symmetric than the original C ones.

Anyway, this PR is a step in the right direction, let us merge it.

**Pull request merged**